libraries/AP_OLC/AP_OLC.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * AP_OLC is based on INAV olc.c implemention, thanks @fiam and other contributors.
+ */
+
+#include <math.h>
+#include <algorithm>
+#include "AP_OLC.h"
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+
+// This is a port of https://github.com/google/open-location-code/blob/master/c/olc.c
+// to avoid double floating point math and use integer math as much as possible.
+
+static constexpr char SEPARATOR_CHAR = '+';
+static constexpr int SEPARATOR_POS = 8;
+static constexpr char PADDING_CHAR = '0';
+
+static constexpr int ENCODING_BASE = 20;
+static constexpr unsigned int PAIR_CODE_LEN = 10;
+static constexpr unsigned int CODE_LEN_MAX = 15;
+
+static constexpr int GRID_COLS = 4;
+static constexpr int GRID_ROWS = ENCODING_BASE / GRID_COLS;
+
+
+static constexpr olc_coord_t LAT_MAX = 90 * OLC_DEG_MULTIPLIER;
+static constexpr olc_coord_t LON_MAX = 180 * OLC_DEG_MULTIPLIER;
+
+olc_coord_t AP_OLC::grid_size;
+olc_coord_t AP_OLC::initial_resolution;
+const char AP_OLC::alphabet[] = "23456789CFGHJMPQRVWX";
+
+bool AP_OLC::inited = false;
+
+
+void AP_OLC::init_constants(void)
+{
+    if (inited) {
+        return;
+    }
+    inited = true;
+
+    // Work out the encoding base exponent necessary to represent 360 degrees.
+    olc_coord_t initial_exponent = floorf(logf(2 * (LON_MAX / OLC_DEG_MULTIPLIER)) / logf(ENCODING_BASE));
```

Comment:
These looks like a constant to me.

They ought to be calculated ahead of time?